RFC#50-P3 Navigation Item Focus (Styling and Behaviour)

Ridhwana opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
In this PR - Part 3 of a multiple series PR, I have focussed predominantly on the sidebar styling and behaviour when one click on it.

## Related Tickets & Documents
https://github.com/forem/rfcs/pull/50

Previous Parts of the PR:
1. https://github.com/forem/forem/pull/12851
2. https://github.com/forem/forem/pull/12886

## QA Instructions, Screenshots, Recordings



### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added tests?

- [ ] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

Ridhwana commented on app/views/admin/shared/_nested_sidebar.html.erb:
```diff
@@ -0,0 +1,29 @@
+<% menu_items.each do |group_name, group| %>
```
I renamed the file from _nested_sidebar.erb to _nested_sidebar.html.erb. and added some classes for current links and making the dropdowns expandable and collapsible.

lisasy replied:
Great change!

Ridhwana commented:
@lisasy I updated the `.crayons-link` to _not_ have text decoration. I know this link is used in other views as well, so I can update it to use another class instead if we want to keep the underline.

In addition, `crayons-link--current` has a font-weight of bold set. However, in the designs the font-weight is normal. Should I keep to this design and create an overriding class or keep it consistent with the rest?

Ridhwana commented on app/assets/stylesheets/components/navigation.scss:
```diff
@@ -2,9 +2,11 @@
 
 .crayons-link {
   color: var(--link-color);
+  text-decoration: none;
```
@lisasy this is the change I'm referring to.

lisasy replied:
@Ridhwana The underline isn't appearing for other places using `.crayons-link`, so I am going to ask @ludwiczakpawel to review this and see why the underline is appearing for these navigation items, when that isn't the case in `/readinglist` or `/notifications`, for example.

My current hypothesis is that `text-decoration: underline` is currently rendering for any `a` item in `/admin` because `/admin` is still pulling in styles from Bootstrap. Eventually, we will deprecate Bootstrap completely so that all UI stylings will be crayons-only. 

If this is the case, then I would recommend using an admin-specific class name with `text-decoration: none` for these `a` elements. When we fully deprecate Bootstrap from Admin Portal, we can also remove these admin-specific class names.

ludwiczakpawel replied:
What @lisasy says is correct - it's coming from bootstrap and it's not an issue on user facing side of DEV because we use bootstrap only in admin section. And as long as we use it, we will likely face some bigger or smaller styling conflicts like this one.

Until we get rid of bootstrap, I highly recommend making "one-off" changes and fixes like that in admin specific stylesheet file: https://github.com/forem/forem/blob/master/app/assets/stylesheets/admin.scss - where you could also leave a comment or something explaining WHY this fix is there so once we get to refactoring this, it will be a bit easier :).

Ridhwana replied:
Sounds good - Thanks you both  😊

lisasy commented:
@Ridhwana Great PR ⚡ There are two interaction design details I'd recommend:


### 1) How a user selects a nav item with nested items
#### Current
Currently, when a user selects "Customization" (for example), it expands all the option, then they must select *another* option. 

#### Proposed
Can we edit the interaction so that when a user selects "Customization", it automatically selects the first child item in nested list — in this case, it would be "Config".

This change helps people drill down to the options they're looking for more easily without asking them to poke around further.

### 2) Showing only one expanded set of nested nav at a time

#### Current
Currently, a user can see multiple nested nav items at a time.

#### Proposed
Because of the proposed change in #1, a user can only see one expanded set of nested nav items at a time.

This change eliminates showing people too many options at the same time.

---

I apologize that I didn't get to explain these details in the design mock. I hope these proposed updates are implementable for you — let me know if you have any questions, thoughts, or concerns.

lisasy reviewed (changes requested):
I'm requesting changes related to these comments:

1. Rolling back on adding `text-decoration: none` to `crayons-link`: https://github.com/forem/forem/pull/12936/files#r589788362
2. Updating interaction design of how nested nav links open: https://github.com/forem/forem/pull/12936#issuecomment-793132956

citizen428 commented on app/lib/menu.rb:
```diff
@@ -13,8 +13,8 @@ def initialize
-  def scope(name, children)
-    @items[name] = children
+  def scope(name, svg, children)
+    @items[name] = { svg: svg, children: children }
```
How about this? If we know all the icons have to be SVGs we can simplify menu creation by adding the file extension here:

```suggestion
    @items[name] = { svg: "#{svg}.svg", children: children }
```

citizen428 commented on app/models/admin_menu.rb:
```diff
@@ -57,20 +57,20 @@ class AdminMenu
-    ITEMS[group_name.to_sym].each do |items|
-      return items if items[:controller] == child_nav_item
+  def self.nested_menu_items(scope_name, nav_item)
+    ITEMS[scope_name.to_sym][:children].each do |items|
```
```suggestion
    ITEMS.dig(scope_name.to_sym, :children).each do |items|
```

citizen428 commented on app/views/admin/shared/_nested_sidebar.html.erb:
```diff
@@ -0,0 +1,29 @@
+<% menu_items.each do |group_name, group| %>
+  <% if group[:children].length == 1 %>
+    <a class="crayons-link crayons-link--block <%= "crayons-link--current" if (deduced_controller(request) == group[:children][0][:controller]) %>"
```
Looking forward to Rails 6.1, where we have `class_names` to simplify this a bit. It will look like this:

```ruby
class_list(
  'crayons-link', 
  'crayons-link--block', 
  { 'crayons-link--current': deduced_controller(request) == group[:children][0][:controller] }
)
```

Ridhwana commented:
> @Ridhwana Great PR ⚡ There are two interaction design details I'd recommend:
> 
> ### 1) How a user selects a nav item with nested items
> #### Current
> Currently, when a user selects "Customization" (for example), it expands all the option, then they must select _another_ option.
> 
> #### Proposed
> Can we edit the interaction so that when a user selects "Customization", it automatically selects the first child item in nested list — in this case, it would be "Config".
> 
> This change helps people drill down to the options they're looking for more easily without asking them to poke around further.
> 
> ### 2) Showing only one expanded set of nested nav at a time
> #### Current
> Currently, a user can see multiple nested nav items at a time.
> 
> #### Proposed
> Because of the proposed change in #1, a user can only see one expanded set of nested nav items at a time.
> 
> This change eliminates showing people too many options at the same time.
> 
> I apologize that I didn't get to explain these details in the design mock. I hope these proposed updates are implementable for you — let me know if you have any questions, thoughts, or concerns.

@lisasy thanks, I've reverted the crayons-link change and amended as per Pawel's suggestion. 

The above request is doable, and I'm currently working on it. It will however require some time and so my suggestion is to get this PR's code reviewed and merged this PR into master (if all the current code looks good) and I'll open a new more focussed PR for these two interaction updates. 

It won't make a difference to the users since it's behind a feature flag.

aitchiss commented on app/assets/stylesheets/admin.scss:
```diff
@@ -188,3 +188,16 @@ label {
   min-width: 120px;
 }
+
+// Navbar links have text-decorations that are being
```
(request / blocking) - Not related to this particular CSS but I notice we don't have any focus styles for these links - making it inaccessible for keyboard users - can we add some here?

Ridhwana replied:
@ludwiczakpawel I'm trying to get the focus styles for admin working. After some investigation I noticed [this css](https://github.com/forem/forem/blob/master/app/assets/stylesheets/scaffolds.scss#L16-L18). I know that you probably didnt add this, but do you by any chance  know why it could have been added ? It seems like if i comment this out, the focus works as expected, but i dont want to break anything unexpected. This gave me some more clues https://css-tricks.com/the-focus-visible-trick/

Ridhwana replied:
> Not related to this particular CSS but I notice we don't have any focus styles for these links - making it inaccessible for keyboard users - can we add some here?
✅
Thanks to Pawel for helping me on this. I've added focus styles on the sidebar specifically - The focus elements on the admin interface as a whole needs to be fixed but I kept this change constrained to the feature I'm working on at the moment to avoid introducing bugs.

aitchiss commented on app/views/admin/shared/_nested_sidebar.html.erb:
```diff
@@ -0,0 +1,29 @@
+      <%= display_name(group_name) %>
+    </a>
+  <% else %>
+    <div class="crayons-link crayons-link--block cursor-pointer <%= "crayons-link--current" if (deduced_scope(request) == group_name.to_s) %>"
+      data-toggle="collapse"
+      data-target="#<%= group_name %>"
+      aria-expanded="<%= (deduced_scope(request) == group_name.to_s).to_s %>"
+      aria-controls="<%= group_name %>">
+      <%= inline_svg_tag("#{group[:svg]}", aria: true, class: "dropdown-icon crayons-icon", title: "#{group_name}") %>
+      <%= display_name(group_name) %>
+    </div>
```
(refactor / blocking) - As this item is a `div` rather than an interactive element, it's not currently possible by keyboard to expand/collapse the menus - can we change to a button? If it helps, there's a nice example of this in the [WAI ARIA practices example](https://www.w3.org/TR/wai-aria-practices/examples/disclosure/disclosure-navigation.html)

It's also best practice to immediately focus the first link in the list after the expanded items are expanded - is it possible to trigger this in the controller when we toggle the list?
 
Nice one with the `aria-expanded` and `controls` 👍

Ridhwana replied:
> can we change to a button? 
✅

>is it possible to trigger this in the controller when we toggle the list?
@lisasy brought this up [here for different reasons](https://github.com/forem/forem/pull/12936#pullrequestreview-606770573). I'm working on a bigger change for the interaction design, and I'll incorporate this change in this follow up PR.

aitchiss commented on app/views/admin/shared/_nested_sidebar.html.erb:
```diff
@@ -0,0 +1,29 @@
+      <%= inline_svg_tag("#{group[:svg]}", aria: true, class: "dropdown-icon crayons-icon", title: "#{group_name}") %>
+      <%= display_name(group_name) %>
+    </div>
+    <div id="<%= group_name %>"
+         class="<%= (deduced_scope(request) == group_name.to_s) ? "expand show" : "collapse hide" %>" aria-labelledby="<%= group_name %>">
+      <% group[:children].each do |item| %>
+        <a class="crayons-link crayons-link--block ml-7 <%= "fw-bold" if deduced_controller(request) == item[:controller].to_s %>"
+          href="/admin/<%= group_name %>/<%= item[:controller] %>"
+        >
+          <%= item[:name].to_s.titleize %>
+        </a>
+      <% end %>
+    </div>
```
(refactor / blocking) - Can we convert this `div` to a `<ul>` and put the individual links inside `<li>` elements? This will let screen reader users know how many items are in the list and announce the current position e.g. 1 of 3.

Same goes for the overall top-level options - Overview/People/etc - these should be in a list too.

For the individual links, can we add `aria-current="page"` to whichever link is the currently active page? Again this just helps screen reader users know where they are as they browse the menu 🙂

aitchiss reviewed (changes requested):
This is looking really nice! I have a few requests for improving the accessibility of this further, although you've already added lots of good stuff - let me know if you have any questions about any of it 🙂

Ridhwana commented:
> This is looking really nice! I have a few requests for improving the accessibility of this further, although you've already added lots of good stuff - let me know if you have any questions about any of it 🙂

@aitchiss thank you for the great review, I learned more about accessibility by addressing your feedback. 🌸 Can you please review again 🙂

lisasy commented:
> > @Ridhwana Great PR ⚡ There are two interaction design details I'd recommend:
> > ### 1) How a user selects a nav item with nested items
> > #### Current
> > Currently, when a user selects "Customization" (for example), it expands all the option, then they must select _another_ option.
> > #### Proposed
> > Can we edit the interaction so that when a user selects "Customization", it automatically selects the first child item in nested list — in this case, it would be "Config".
> > This change helps people drill down to the options they're looking for more easily without asking them to poke around further.
> > ### 2) Showing only one expanded set of nested nav at a time
> > #### Current
> > Currently, a user can see multiple nested nav items at a time.
> > #### Proposed
> > Because of the proposed change in #1, a user can only see one expanded set of nested nav items at a time.
> > This change eliminates showing people too many options at the same time.
> > I apologize that I didn't get to explain these details in the design mock. I hope these proposed updates are implementable for you — let me know if you have any questions, thoughts, or concerns.
> 
> @lisasy thanks, I've reverted the crayons-link change and amended as per Pawel's suggestion.
> 
> The above request is doable, and I'm currently working on it. It will however require some time and so my preference is to get this PR's code reviewed and merged this PR into master and I'll open a new focussed PR for these two interaction updates.
> 
> It won't make a difference to the users since it's behind a feature flag.

Sounds good — thanks. I'll approve this :)

jgaskins commented on config/routes.rb:
```diff
@@ -219,7 +219,7 @@
         end
 
-        scope path: :app do
+        scope path: :apps do
```
Do we need to account for any existing links to `app`?

Ridhwana replied:
@jgaskins no this was actually a typo which led to a bug. However, because it was under a feature flag and a WIP it did not affect any users and just needed to be rectified as we make progress up this feature.

jgaskins commented on app/views/admin/shared/_nested_sidebar.html.erb:
```diff
@@ -0,0 +1,36 @@
+        aria-expanded="<%= (deduced_scope(request) == group_name.to_s).to_s %>"
+        aria-controls="<%= group_name %>"
+      >
+        <%= inline_svg_tag("#{group[:svg]}", aria: true, class: "dropdown-icon crayons-icon", title: "#{group_name}") %>
```
Minor nitpick 😄 

```suggestion
        <%= inline_svg_tag(group[:svg], aria: true, class: "dropdown-icon crayons-icon", title: group_name) %>
```

Ridhwana replied:
Thanks for picking this one up @jgaskins! 😊 I removed the .svg into the menu file and forgot to remove the string interpolation ✅

jgaskins commented on app/views/admin/shared/_nested_sidebar.html.erb:
```diff
@@ -0,0 +1,36 @@
+              href="/admin/<%= group_name %>/<%= item[:controller] %>"
+              aria-page="<%= "page" if deduced_controller(request) == item[:controller].to_s %>"
+            >
+              <%= item[:name].to_s.titleize %>
```
I may be looking too hard for patterns here, but the single-item special case above uses the `display_name` helper method. Should we use that here, as well?

```suggestion
              <%= display_name(item[:name]) %>
```

Ridhwana replied:
@jgaskins no it wouldn't and here's the reason:

`group_name` is the name of a group like `:content_manager` - it's the first outer key in the payload, hence we need to manipulate it using `display_name`  by replacing underscores with spaces and using titleize on it. 

whereas the `item[:name]` is the keys inside the payload for the groups navigation items. Since we do not always want to use the controller_names for these items, we specify names for the navigation items in the payload which already include the spaces. Hence we just make sure it's titleized. Hope this makes sense, feel free to ask any follow up questions

jgaskins reviewed (commented):
Just some questions as I continue to get familiar with the codebase

Ridhwana commented on app/assets/stylesheets/admin.scss:
```diff
@@ -188,3 +188,35 @@ label {
+    padding: 0;
+  }
+
+  :focus-visible {
+    outline: 2px auto blue;
+  }
+
+  :focus:not(:focus-visible) {
+    outline: none;
+  }
```
It seems focus-visible is not supported in Safari. Looking for a workaround.

aitchiss replied:
looks like this should work:

```scss
.crayons-link {
    text-decoration: none;

    &:focus {
      outline: 2px auto blue;
    }

    &:hover {
      text-decoration: none;
    }
  }
```

Ridhwana replied:
This article was great in helping me to figure out the issue. https://matthiasott.com/notes/focus-visible-is-here: Using progressive enhancement we can define regular focus styles for non-supporting browsers and then overwrite them for browsers that support :focus-visible.

aitchiss replied:
or rather I mean move all the focus styles into here, to avoid seeing that focus indicator on click too - 

```scss
.crayons-link {
    text-decoration: none;

    &:focus {
      outline: 2px auto blue;
    }

    &:hover {
      text-decoration: none;
    }

    &:focus:not(:focus-visible) {
      outline: none;
    }
  }
```

Ridhwana replied:
Thanks @aitchiss! Done and pushed ✅ 😊

aitchiss reviewed (approved):
Nice one! This looks great - tested locally on Safari/Firefox/Chrome and with VoiceOver on Safari and all looks good to me 🚀
